File: board.cc
#include "board.h"

#include <cstring>
#include <glog/logging.h>

#define ANSI_COLOR_RESET   "\x1b[0m"

namespace blokus {

namespace {

// Returns all translated, rotated, and flipped coordinates for the placed
// tile.
// TODO(piotrf): consider exposing this as a public api.
std::vector<Coord> PlacedTile(const Tile& tile, const Move& move) {
  std::vector<Coord> coors;
  for (int i = 0; i < 5; ++i) {
    for (int j = 0; j < 5; ++j) {
      if (tile[i][j]) {
        Coord c;
        switch (move.rotation) {
          case 0: c[0] = i; c[1] = j; break;
          case 1: c[0] = -j; c[1] = i; break;
          case 2: c[0] = -i; c[1] = -j; break;
          case 3: c[0] = j; c[1] = -i; break;
          default:
            LOG(FATAL) << "Unknown rotation: " << move.rotation;
        }
        if (move.flip) {
          c[0] = -c[0];
        }
        coors.push_back(std::move(c));
      }
    }
  }
  for (auto& coor : coors) {
    coor[0] += move.coord[0];
    coor[1] += move.coord[1];
  }
  return coors;
}

}  // namespace

std::string Move::DebugString() const {
  char output[12];
  snprintf(output, 12, "(%2d,%2d) %d %d", coord[0], coord[1], rotation, flip);
  return output;
}

std::string ColorToString(Color c) {
  switch(c) {
    case BLUE: return "Blue";
    case YELLOW: return "Yellow";
    case RED: return "Red";
    case GREEN: return "Green";
    default:
      LOG(FATAL) << "Unknown color: " << c;
  }
}

std::string AnsiColor(Color c) {
  switch(c) {
    case BLUE: return "\x1b[34m";
    case YELLOW: return "\x1b[33m";
    case RED: return "\x1b[31m";
    case GREEN: return "\x1b[32m";
    default:
      LOG(FATAL) << "Unknown color: " << c;
  }
}

Board::Board() {
  // Initially, there are no pieces on the board.
  memset(pieces_, 0, kWidth * kHeight * sizeof(uint8_t));

  // Initially, you can only move in a corner. Place in turn order,
  // going clockwise from top-left (0,0).
  memset(frontier_, 0, kWidth * kHeight * sizeof(uint8_t));
  frontier_[0][0] = BLUE;
  frontier_[kWidth - 1][0] = YELLOW;
  frontier_[kWidth - 1][kHeight - 1] = RED;
  frontier_[0][kHeight - 1] = GREEN;

  // Initially, everyone is allowed to move everywhere.
  memset(allowed_, BLUE | YELLOW | RED | GREEN,
         kWidth * kHeight * sizeof(uint8_t));
}

Board::~Board() {
}

// Returns true if it is possible to place the given tile, of the given color,
// in the given move, respecting geometry of the board and rules of the game.
bool Board::IsPossible(const Tile& tile, Color color, const Move& move) const {
  bool contains_frontier = false;
  for (const Coord& coord : PlacedTile(tile, move)) {
    int x = coord[0];
    int y = coord[1];

    // Verify that the piece is on the board.
    if (x < 0 || x > kWidth - 1) return false;
    if (y < 0 || y > kHeight - 1) return false;

    // Verify that we are allowed to move here.
    if ((allowed_[x][y] & color) == 0) return false;

    // Keep track of whether or not we touch the frontier. We must touch in at
    // least on place.
    contains_frontier |= frontier_[coord[0]][coord[1]] & color;
  }
  return contains_frontier;
}

// Returns a list of all possible moves for the given tile and color.
// TODO(piotrf): currently, this can list duplicates of essentially the same
// move, since the resulting board state may be the same for different
// coordinates and a different rotation.
std::vector<Move> Board::PossibleMoves(const Tile& tile, Color color) const {
  std::vector<Move> moves;

  // Naive, brute-force algorithm: try all possible places on the board, with
  // all rotations and flips.
  // TODO(piotrf): implement a more efficient search that ignores things that
  // obviously don't work.
  for (int x = 0; x < kWidth; ++x) {
    for (int y = 0; y < kHeight; ++y) {
      for (int rotation = 0; rotation < 4; rotation++) {
        for (bool flip : {false, true}) {
          Move move;
          move.coord = Coord(x, y);
          move.rotation = rotation;
          move.flip = flip;
          if (IsPossible(tile, color, move)) {
            moves.push_back(std::move(move));
          }
        }
      }
    }
  }

  return moves;
}

bool Board::Place(const Tile& tile, Color color, const Move& move) {
  if (!IsPossible(tile, color, move)) {
    return false;
  }
  std::vector<Coord> coords = PlacedTile(tile, move);
  for (const Coord& coord : coords) {
    pieces_[coord[0]][coord[1]] = color;
  }
  // Update frontier and allowed based on current state of pieces_.
  for (int i = 0; i < kWidth; ++i) {
    for (int j = 0; j < kHeight; ++j) {
      // If there is a piece here, it's not on the frontier or allowed.
      if (pieces_[i][j]) {
        frontier_[i][j] = 0;
        allowed_[i][j] = 0;
        continue;
      }
      // If there is a piece horizontally or vertically separated, then
      // that color can't move here.
      uint8_t allowed = BLUE | YELLOW | RED | GREEN;
      if (i > 0 && pieces_[i - 1][j]) allowed &= ~pieces_[i - 1][j];
      if (i < kWidth - 1 && pieces_[i + 1][j]) allowed &= ~pieces_[i + 1][j];
      if (j > 0 && pieces_[i][j - 1]) allowed &= ~pieces_[i][j - 1];
      if (j < kHeight - 1 && pieces_[i][j + 1]) allowed &= ~pieces_[i][j + 1];
      allowed_[i][j] = allowed;
      // If there is a piece corner separated from us, and this is an allowed
      // spot, then this is on our frontier.
      uint8_t frontier = 0;
      if (i > 0 && j > 0 && pieces_[i - 1][j - 1])
        frontier |= pieces_[i - 1][j - 1];
      if (i > 0 && j < kHeight - 1 && pieces_[i - 1][j + 1])
        frontier |= pieces_[i - 1][j + 1];
      if (i < kWidth - 1 && j > 0 && pieces_[i + 1][j - 1])
        frontier |= pieces_[i + 1][j - 1];
      if (i < kWidth - 1 && j < kHeight - 1 && pieces_[i + 1][j + 1])
        frontier |= pieces_[i + 1][j + 1];
      frontier_[i][j] = frontier & allowed;
    }
  }
  // Always enforce initial frontier.
  // TODO(piotrf): clean up code duplication.
  frontier_[0][0] = BLUE;
  frontier_[kWidth - 1][0] = YELLOW;
  frontier_[kWidth - 1][kHeight - 1] = RED;
  frontier_[0][kHeight - 1] = GREEN;
  return true;
}

void Board::PrintAllowable(const uint8_t board[kWidth][kHeight]) const {
  printf("  ");
  for (int i = 0; i < kWidth; ++i) {
    printf("  %2d ", i);
  }
  printf("\n");
  for (int j = 0; j < kHeight; ++j) {
    printf("%2d ", j);
    for (int i = 0; i < kWidth; ++i) {
      std::string out;
      for (Color color : {BLUE, YELLOW, RED, GREEN}) {
        if (board[i][j] & color)
          out += AnsiColor(color) + "\u25a3";
        else
          out +=  " ";
      }
      out += ANSI_COLOR_RESET " ";
      printf("%s", out.c_str());
    }
    printf("\n");
  }
}

void Board::Print(bool debug) const {
  // Print out the pieces on the board.
  printf("  ");
  for (int i = 0; i < kWidth; ++i) {
    printf(" %2d", i);
  }
  printf("\n");
  for (int j = 0; j < kHeight; ++j) {
    printf("%2d ", j);
    for (int i = 0; i < kWidth; ++i) {
      if (pieces_[i][j]) {
        std::string out =
            AnsiColor(static_cast<Color>(pieces_[i][j])) + "\u25a3";
        printf(" %s" ANSI_COLOR_RESET " ", out.c_str());
      } else {
        printf(" _ ");
      }
    }
    printf("\n");
  }
  if (debug) {
    printf("========== FRONTIER ==========\n");
    PrintAllowable(frontier_);
    printf("========== ALLOWED ==========\n");
    PrintAllowable(allowed_);
  }
}

}  // namespace blokus

File: board.js
// Lots of algorithms are shared between here and board.cc :/
var board = board || {};

board.WIDTH = 20;
board.HEIGHT = 20;

// Enum for player values.
board.BLUE = 1;
board.YELLOW = 2;
board.RED = 4;
board.GREEN = 8;

board.playerColor = function(player) {
  switch (player) {
    case board.BLUE: return "rgb(0, 0, 200)";
    case board.YELLOW: return "rgb(200, 200, 0)";
    case board.RED: return "rgb(200, 0, 0)";
    case board.GREEN: return "rgb(0, 200, 0)";
  }
}

board.playerString = function(player) {
  switch (player) {
    case board.BLUE: return "blue";
    case board.YELLOW: return "yellow";
    case board.RED: return "red";
    case board.GREEN: return "green";
  }
}

board.nextPlayer = function(player) {
  switch (player) {
    case board.BLUE: return board.YELLOW;
    case board.YELLOW: return board.RED;
    case board.RED: return board.GREEN;
    case board.GREEN: return board.BLUE;
  }
}

function Create2DArray(width, height) {
  var array = Array(width);
  for (var i = 0; i < width; ++i) {
    array[i] = Array(height);
  }
  return array;
}

board.Board = function() {
  // Initially, there are no pieces on the board.
  this.pieces = Create2DArray(board.WIDTH, board.HEIGHT);

  // Initially, you can only move in a corner. Place in turn order,
  // going clockwise from top-left (0,0).
  this.frontier = Create2DArray(board.WIDTH, board.HEIGHT);
  this.frontier[0][0] = board.BLUE;
  this.frontier[board.WIDTH - 1][0] = board.YELLOW;
  this.frontier[board.WIDTH - 1][board.HEIGHT - 1] = board.RED;
  this.frontier[0][board.HEIGHT - 1] = board.GREEN;

  // Initially, everyone is allowed to move everywhere.
  this.allowed = Create2DArray(board.WIDTH, board.HEIGHT);
  for (var i = 0; i < board.WIDTH; ++i) {
    for (var j = 0; j < board.HEIGHT; ++j) {
      this.allowed[i][j] = board.BLUE | board.YELLOW | board.RED | board.GREEN;
    }
  }
}

board.Board.prototype.isPossible = function(tile, color, move) {
  var containsFrontier = false;
  for (var i = 0; i < tile.coors.length; ++i) {
    // Transform the tile coordinates based on rotation and flip.
    // TODO(piotrf): fix code duplication between here and tiles.js
    var tX, tY;
    switch(move.rotation) {
      case 0:
        tX = tile.coors[i][0]
        tY = tile.coors[i][1];
        break;
      case 1:
        tX = -tile.coors[i][1];
        tY = tile.coors[i][0];
        break;
      case 2:
        tX = -tile.coors[i][0];
        tY = -tile.coors[i][1];
        break;
      case 3:
        tX = tile.coors[i][1];
        tY = -tile.coors[i][0];
        break;
    }
    if (move.flip) {
      tX = -tX;
    }
    var x = tX + move.coord[0];
    var y = tY + move.coord[1];
    
    // Verify that the piece is on the board.
    if (x < 0 || x > board.WIDTH - 1) return false;
    if (y < 0 || y > board.HEIGHT - 1) return false;

    // Verify that we are allowed to move here.
    if ((this.allowed[x][y] & color) == 0) return false;

    // Keep track of whether or not we touch the frontier. We must touch in at
    // least one place.
    containsFrontier |= this.frontier[x][y] & color;
  }
  return containsFrontier;
}

board.Board.prototype.place = function(tile, color, move) {
  if (!this.isPossible(tile, color, move)) {
    alert('Error! Trying to place in an impossible location.');
    return;
  }
  // Transform the tile coordinates based on rotation and flip.
  // TODO(piotrf): fix code duplication between here and tiles.js
  for (var i = 0; i < tile.coors.length; ++i) {
    var tX, tY;
    switch(move.rotation) {
      case 0:
        tX = tile.coors[i][0]
        tY = tile.coors[i][1];
        break;
      case 1:
        tX = -tile.coors[i][1];
        tY = tile.coors[i][0];
        break;
      case 2:
        tX = -tile.coors[i][0];
        tY = -tile.coors[i][1];
        break;
      case 3:
        tX = tile.coors[i][1];
        tY = -tile.coors[i][0];
        break;
    }
    if (move.flip) {
      tX = -tX;
    }
    var x = tX + move.coord[0];
    var y = tY + move.coord[1];
    this.pieces[x][y] = color;
  }
  // Update frontier and allowed based on current state of pieces.
  for (var i = 0; i < board.WIDTH; ++i) {
    for (var j = 0; j < board.HEIGHT; ++j) {
      // If there is a piece here, it's not on the frontier or allowed.
      if (this.pieces[i][j]) {
        this.frontier[i][j] = 0;
        this.allowed[i][j] = 0;
        continue;
      }
      // If there is a piece horizontally or vertically separated, then
      // that color can't move here.
      var allowed = board.BLUE | board.YELLOW | board.RED | board.GREEN;
      if (i > 0 && this.pieces[i - 1][j])
	allowed &= ~this.pieces[i - 1][j];
      if (i < board.WIDTH - 1 && this.pieces[i + 1][j])
	allowed &= ~this.pieces[i + 1][j];
      if (j > 0 && this.pieces[i][j - 1])
	allowed &= ~this.pieces[i][j - 1];
      if (j < board.HEIGHT - 1 && this.pieces[i][j + 1])
	allowed &= ~this.pieces[i][j + 1];
      this.allowed[i][j] = allowed;
      // If there is a piece corner separated from us, and this is an allowed
      // spot, then this is on our frontier.
      var frontier = 0;
      if (i > 0 && j > 0 && this.pieces[i - 1][j - 1])
        frontier |= this.pieces[i - 1][j - 1];
      if (i > 0 && j < board.HEIGHT - 1 && this.pieces[i - 1][j + 1])
        frontier |= this.pieces[i - 1][j + 1];
      if (i < board.WIDTH - 1 && j > 0 && this.pieces[i + 1][j - 1])
        frontier |= this.pieces[i + 1][j - 1];
      if (i < board.WIDTH - 1 && j < board.HEIGHT - 1 &&
	  this.pieces[i + 1][j + 1])
        frontier |= this.pieces[i + 1][j + 1];
      this.frontier[i][j] = frontier & allowed;

    }
  }
  // Always enforce initial frontier.
  this.frontier[0][0] = board.BLUE;
  this.frontier[board.WIDTH - 1][0] = board.YELLOW;
  this.frontier[board.WIDTH - 1][board.HEIGHT - 1] = board.RED;
  this.frontier[0][board.HEIGHT - 1] = board.GREEN;
}

board.Board.prototype.draw = function(ctx) {
  for (var i = 0; i < board.WIDTH; ++i) {
    for (var j = 0; j < board.HEIGHT; ++j) {
      if (this.pieces[i][j] > 0) {
	ctx.fillStyle = board.playerColor(this.pieces[i][j])
      } else {
        ctx.fillStyle = "rgb(200, 200, 200)";
      }
      ctx.fillRect(BOARD_SCALE * i + 1, BOARD_SCALE * j + 1, BOARD_SCALE - 2,
		   BOARD_SCALE - 2);
    }
  }
}

